File: java/src/Exmo.java
/**
 * Created by Admin on 2/18/2016.
 */

import okhttp3.*;
import org.apache.commons.codec.binary.Hex;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.HashMap;
import java.util.Map;


public class Exmo {
    private static long _nonce;
    private String _key;
    private String _secret;

    public Exmo(String key, String secret) {
        _nonce = System.currentTimeMillis();
        _key = key;
        _secret = secret;
    }

    public final String Request(String method, Map<String, String> arguments) {
        Mac mac;
        SecretKeySpec key;
        String sign;

        if (arguments == null) {  // If the user provided no arguments, just create an empty argument array.
            arguments = new HashMap<>();
        }

        arguments.put("nonce", "" + ++_nonce);  // Add the dummy nonce.

        String postData = "";

        for (Map.Entry<String, String> stringStringEntry : arguments.entrySet()) {
            Map.Entry argument = (Map.Entry) stringStringEntry;

            if (postData.length() > 0) {
                postData += "&";
            }
            postData += argument.getKey() + "=" + argument.getValue();
        }

        // Create a new secret key
        try {
            key = new SecretKeySpec(_secret.getBytes("UTF-8"), "HmacSHA512");
        } catch (UnsupportedEncodingException uee) {
            System.err.println("Unsupported encoding exception: " + uee.toString());
            return null;
        }

        // Create a new mac
        try {
            mac = Mac.getInstance("HmacSHA512");
        } catch (NoSuchAlgorithmException nsae) {
            System.err.println("No such algorithm exception: " + nsae.toString());
            return null;
        }

        // Init mac with key.
        try {
            mac.init(key);
        } catch (InvalidKeyException ike) {
            System.err.println("Invalid key exception: " + ike.toString());
            return null;
        }


        // Encode the post data by the secret and encode the result as base64.
        try {
            sign = Hex.encodeHexString(mac.doFinal(postData.getBytes("UTF-8")));
        } catch (UnsupportedEncodingException uee) {
            System.err.println("Unsupported encoding exception: " + uee.toString());
            return null;
        }

        // Now do the actual request
        MediaType form = MediaType.parse("application/x-www-form-urlencoded; charset=utf-8");

        OkHttpClient client = new OkHttpClient();
        try {

            RequestBody body = RequestBody.create(form, postData);
            Request request = new Request.Builder()
                    .url("https://api.exmo.com/v1/" + method)
                    .addHeader("Key", _key)
                    .addHeader("Sign", sign)
                    .post(body)
                    .build();
            
            Response response = client.newCall(request).execute();
            return response.body().string();
        } catch (IOException e) {
            System.err.println("Request fail: " + e.toString());
            return null;  // An error occured...
        }
    }
}

File: nodejs/exmo.js
var crypto = require('crypto')
	http = require('http'),
	querystring = require('querystring'),
	request = require('request'),
	config = {
		url: 'https://api.exmo.com/v1/'
	};
	

function sign(message){
    return crypto.createHmac('sha512', config.secret).update(message).digest('hex');
}

exports.init_exmo = function (cfg) {
	config.key = cfg.key;
	config.secret = cfg.secret;
	config.nonce = Math.floor(new Date().getTime());
};

exports.api_query = function(method_name, data, callback){
	data.nonce = config.nonce++;
	var post_data = querystring.stringify(data);

	var options = {
	  url: config.url + method_name,
	  method: 'POST',
	  headers: {
	    'Key': config.key,
	    'Sign': sign(post_data)
	  },
	  form:data
	};
	
 	request(options, function (error, response, body) {
        if (!error && response.statusCode == 200) {
            callback(body);
        }else{
        	callback(error);
        }
    });
};

exports.api_query2 = function(method_name, data, callback){
	data.nonce = config.nonce++;
	var post_data = querystring.stringify(data);

	var post_options = {
	  	host: 'api.exmo.com',
	  	port: '80',
	  	path: '/v1/' + method_name,
	  	method: 'POST',
		headers: {
			'Key': config.key,
			'Sign': sign(post_data),
			'Content-Type': 'application/x-www-form-urlencoded',
			'Content-Length': Buffer.byteLength(post_data)
		}
	};
 	var post_req = http.request(post_options, function(res) {
      res.setEncoding('utf8');
      res.on('data', function (chunk) {
          callback(chunk);
      });
  	});

  	post_req.write(post_data);
  	post_req.end();
};



exports.test = function(){
    return config.key;
};
